code/modules/power/gravitygenerator.dm

```diff
@@ -144,123 +158,120 @@ GLOBAL_LIST_EMPTY(gravity_generators) // We will keep track of this by adding ne
+	if(construction_state != GRAV_NEEDS_WRENCH)
 		return
+	. = TRUE
+	if(!I.use_tool(src, user, volume = I.tool_volume))
+		return
+	to_chat(user, "<span class='notice'>You secure the plating to the framework.</span>")
+	construction_state = GRAV_NEEDS_SCREWDRIVER
+	update_icon()
 
-	var/dat = "Gravity Generator Breaker: "
-	if(breaker)
-		dat += "<span class='linkOn'>ON</span> <A href='?src=[UID()];gentoggle=1'>OFF</A>"
-	else
-		dat += "<A href='?src=[UID()];gentoggle=1'>ON</A> <span class='linkOn'>OFF</span> "
-
-	dat += "<br>Generator Status:<br><div class='statusDisplay'>"
-	if(charging_state != GRAV_POWER_IDLE)
-		dat += "<font class='bad'>WARNING</font> Radiation Detected. <br>[charging_state == GRAV_POWER_UP ? "Charging..." : "Discharging..."]"
-	else if(on)
-		dat += "Powered."
-	else
-		dat += "Unpowered."
-
-	dat += "<br>Gravity Charge: [charge_count]%</div>"
-
-	var/datum/browser/popup = new(user, "gravgen", name)
-	popup.set_content(dat)
-	popup.open()
-
+// Step 4
+/obj/machinery/gravity_generator/main/screwdriver_act(mob/living/user, obj/item/I)
+	if(!(stat & BROKEN)) // Not actually broken
+		return
+	if(construction_state != GRAV_NEEDS_SCREWDRIVER)
+		return
+	. = TRUE
+	if(!I.use_tool(src, user, volume = I.tool_volume))
+		return
+	to_chat(user, "<span class='notice'>You screw the covers back into place.</span>")
+	set_fix()
 
-/obj/machinery/gravity_generator/main/Topic(href, href_list)
+/obj/machinery/gravity_generator/main/attack_hand(mob/user)
+	if(!..())
+		return ui_interact(user)
+
+/obj/machinery/gravity_generator/main/attack_ai(mob/user)
+	return TRUE
+
+/obj/machinery/gravity_generator/main/attack_ghost(mob/user)
+	return ui_interact(user)
+
+// tgui\packages\tgui\interfaces\GravityGen.js
+/obj/machinery/gravity_generator/main/ui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = FALSE, datum/tgui/master_ui = null, datum/ui_state/state = GLOB.default_state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui && !(stat & BROKEN))
+		ui = new(user, src, ui_key, "GravityGen", name, 350, 350, master_ui, state)
+		ui.open()
+
+/obj/machinery/gravity_generator/main/ui_data(mob/user)
+	var/list/data = list()
+	data["charging_state"] = charging_state
+	data["charge_count"] = charge_count
+	data["breaker"] = breaker_on
+	data["ext_power"] = on
+	return data
+
+/obj/machinery/gravity_generator/main/ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
 	if(..())
-		return 1
+		return
 
-	if(href_list["gentoggle"])
-		breaker = !breaker
-		investigate_log("was toggled [breaker ? "<font color='green'>ON</font>" : "<font color='red'>OFF</font>"] by [usr.key].", "gravity")
+	. = TRUE
+	if(action == "breaker")
+		breaker_on = !breaker_on
+		investigate_log("was toggled [breaker_on ? "<font color='green'>ON</font>" : "<font color='red'>OFF</font>"] by [usr.key].", "gravity")
 		set_power()
-		src.updateUsrDialog()
+		updateUsrDialog()
```

Comment:
This isn't needed, because we don't do any `set_machine()` stuff anymore.
```suggestion
```

Reply:
No idea what these actually do so I'll take your word for it.